File: Java/HaystackNeedle.java
package Java;

public class HaystackNeedle {
  private static int findNeedleInHaystack(String haystack, String needle) {
    int haystackLength = haystack.length(); // 7
    int needleLength = needle.length(); // 4

    for (int i = 0; i <= haystackLength - needleLength; i++) {
      if (haystack.substring(i, needleLength + i).equals(needle)) return i;
    }
    return -1;
  }

  public static void main(String[] args) {
    System.out.println(findNeedleInHaystack("garrafa", "rafa"));
  }
}

/*
 * Recebe haystack e needle, duas strings, e tem que achar a posição em que needle está no haystack
 * guarda o comprimento das duas strings em duas vargs ...Length
 * loop percorre o comprimento da haystack enquanto a needle couber
 *    ou seja, se haystack tem 7 letras, e needle tem 4, vai percorrer 3 posições
 * compara a substring de haystack formada da posição atual até o comprimento de needle + i
 *    o mais i é pq essa posição final deve mudar a cada iteração
 *    tem que ser o tamanho da needle contando a partir da posição i
 */

File: TS/haystackNeedle.ts
function findNeedleInHaystack(haystack: string, needle: string): number {
  if (haystack.length < needle.length) return -1;

  const haystackLength = haystack.length;
  const needleLength = needle.length;

  for (let i = 0; i <= haystackLength - needleLength; i++) {
    if (haystack.substring(i, needleLength + i) === needle) return i;
  }
  return -1;
}

console.log(findNeedleInHaystack("leetcode", "code"));
console.log(findNeedleInHaystack("garrafa", "rafa"));


/*
 * guardar o comprimento das duas strings
 * a needle tem que caber na haystack
 * só pode percorrer a haystack até a última posição em que a needle ainda caiba
 *  ex: haystack 7 letras, needle 3 letras, só pode percorrer até a quarta letra (index 3)
 *  isso se resolve percorrendo enquanto i < tamanho da haystack - tamanho da needle
 * 
 * capturar substring formada pelo indice atual até o comprimento da needle + indice atual
 *  somar comprimento da needle com indice atual pra ajustar a posição em que estamos procurando a needle
 * se a substring for igual a needle, retornar o indice
 * se o loop não encontrar, retornar -1
 */